docs/reddit-migration/strategy/schema.md

```diff
@@ -0,0 +1,61 @@
+# Establish a New Schema
+
+We want to establish a new schema so that we have a clean slate to work with and because there will be some incompatible changes with the current schema. An example of this is the `Post.post_id` column is a reference to the reddit id, but in the future needs to be the primary key of the row (see [Reddit Ids](reddit-ids.md)). We can't do that in-place with the current table.
+### Channels
+
+Channels should have the fields carried over with the following adjustments:
+
+ - `channel_id` -> `id`
+
+### Posts
+
+Posts should have the fields carried over with the following adjustments:
+
+ - `post_id` -> `id`
+ - `preview_text` / `plain_text` - we should to normalize this property / attribute down to a single field so we're not recomputing it every time a read occurs
+
+
+### Comments
+
+Comments should have the fields carried over with the following adjustments:
+
+ - `channel_id` -> `id`
+
+#### Comment Tree
+
+Prior art:
+
+ - SQL Antipatterns ("Naive Trees"):http://www.r-5.org/files/books/computers/languages/sql/style/Bill_Karwin-SQL_Antipatterns-EN.pdf
+ - Models for Hierarchical Data: https://www.slideshare.net/billkarwin/models-for-hierarchical-data
+ - Django treebeard: https://django-treebeard.readthedocs.io/en/latest/
+
+Reddit stores comment trees in Cassandra, which is overkill for our current use cases. That said, we need to be able to migrate the comments in such a manner that we achieve the following:
+
+- Maintain current API result structures
+  - Specifically, we need to support some form of reddit's "more comments" pagination. This **does not** need to partition the comments exactly as reddit would, the interface just needs to be compatible with how the frontend traverses it.
```

Comment:
is there a way we can determine how many posts currently on the production instance actually pass the "more comments" barrier? just wondering if we may be able to get away with leaving out this detail from an MVP for migrating away from reddit.

meaning, if we do (as I suspect) have very very few posts with more than, say, 20 comments we may be falling into a bit of a YAGNI trap if we add comment pagination to the tree initially. Of course we do want to make sure we use a solution that is going to be compatible with it when we do implement it.

Reply:
I think the problem is that if we want to limit the api response to a reasonable size we need to pagination it somehow. I can see 2 ways of doing that:

- Select the first N items from the tree and return "more" links as appropriate
  - This is easy to do with the way `treebeard` structures the table
  - This is reasonably close, but not exactly how reddit does it. I think reddit does some kind of branch clipping based on the relevance/popularity of tree nodes, but we don't need anything nearly that fancy
  - This would mean that we would have 0-2 "more" links in the response:
    - 0 if we loaded the entire tree in 1 response
    - 1 if we cleanly paginated between 2 top-level comments
    - 2 if we paginated in the middle of a branch - 1 in the middle of the branch and another at the top level
- Select *at least* the first N comments, but only paginate between top-level comments
  - The query here might be a little tricky, because you don't really know how many top-level comments you need to select. Might require 2 queries - first one to get N number of comments, second one to get the rest of the branch of the last comment in that query response. Haven't thought too much about this kind of approach though.
  - There's no upper bound on how many items this response could return since the branches are unbounded, although in practice it might not be too bad.
  - Although if conversational engagement on the app gets high enough, we could see a performance hit here

So we could simplify it, but at the risk of having to reintroduce this mechanism when engagement goes up.

Reply:
that makes sense to me, especially given that treebeard has some support for doing this